src/java/org/apache/cassandra/service/StorageService.java

```diff
@@ -868,9 +869,38 @@ private void joinTokenRing(int delay) throws ConfigurationException
             setMode(Mode.WAITING_TO_BOOTSTRAP, "Awaiting start bootstrap call", true);
             bootstrapManager.awaitBootstrappable();
             if (SystemKeyspace.bootstrapInProgress())
-                logger.warn("Detected previous bootstrap failure; retrying");
+            {
+                // on detection of previous bootstrap failure prevent boostrap from proceeding if data directories are
+                // filled up beyond palantir_cassandra.bootstrap_disk_usage_threshold_percentage
+                Integer threshold = Integer.getInteger("palantir_cassandra.bootstrap_disk_usage_threshold_percentage");
+                Entry<Directories.DataDirectory, Double> mostUtilizedDataDir = Directories.getMostUtilizedDataDir();
+                Double percentageDataDirUtilization = mostUtilizedDataDir.getValue() * 100;
+
+                if (threshold != null && percentageDataDirUtilization > Double.valueOf(threshold))
+                {
+                    // disable node preventing bootstrap continuation.
+                    setMode(Mode.NON_TRANSIENT_ERROR, "Detected previous bootstrap failure, data dir too full to proceed.", true);
```

Comment:
Rather than setting the mode directly, let's call `recordNonTransientError` with appropriate args for this type of failure.  That way, we will be able to expose the non transient error via JMX for our health check infrastructure.